drivers/sensors/pmw3320.c

```diff
@@ -0,0 +1,191 @@
+    pmw3320_write_reg(REG_Rest_Mode_Status, 0x00);
+    // Set LED to be always on
+    pmw3320_write_reg(REG_Led_Control, 0x4);
+    // Disable rest mode
+    pmw3320_write_reg(REG_Performance, 0x80);
+}
+
+// Perform a synchronization with sensor.
+// Just as with the serial protocol, this is used by the slave to send a
+// synchronization signal to the master.
+void pmw3320_sync(void) {
+    writePinLow(PMW3320_CS_PIN);
+    wait_us(1);
+    writePinHigh(PMW3320_CS_PIN);
+}
+
+void pmw3320_cs_select(void) {
+    writePinLow(PMW3320_CS_PIN);
+}
+
+void pmw3320_cs_deselect(void) {
+    writePinHigh(PMW3320_CS_PIN);
+}
+
+uint8_t pmw3320_serial_read(void) {
+    setPinInput(PMW3320_SDIO_PIN);
+    uint8_t byte = 0;
+
+    for (uint8_t i = 0; i < 8; ++i) {
+        writePinLow(PMW3320_SCLK_PIN);
+        wait_us(1);
+
+        byte = (byte << 1) | readPin(PMW3320_SDIO_PIN);
+
+        writePinHigh(PMW3320_SCLK_PIN);
+        wait_us(1);
+    }
+
+    return byte;
+}
+
+void pmw3320_serial_write(uint8_t data) {
+    setPinOutput(PMW3320_SDIO_PIN);
+
+    for (int8_t b = 7; b >= 0; b--) {
+        writePinLow(PMW3320_SCLK_PIN);
+
+        if (data & (1 << b))
+            writePinHigh(PMW3320_SDIO_PIN);
+        else
+            writePinLow(PMW3320_SDIO_PIN);
+
+        wait_us(2);
+
+        writePinHigh(PMW3320_SCLK_PIN);
+    }
+
+    // This was taken from ADNS5050 driver.
+    // There's no any info in PMW3320 datasheet about this...
+    // tSWR. See page 15 of the ADNS5050 spec sheet.
+    // Technically, this is only necessary if the next operation is an SDIO
+    // read. This is not guaranteed to be the case, but we're being lazy.
+    wait_us(4);
+
+    // Note that tSWW is never necessary. All write operations require at
+    // least 32us, which exceeds tSWW, so there's never a need to wait for it.
+}
+
+// Read a byte of data from a register on the sensor.
+uint8_t pmw3320_read_reg(uint8_t reg_addr) {
+    pmw3320_cs_select();
+
+    pmw3320_serial_write(reg_addr);
+
+    uint8_t byte = pmw3320_serial_read();
+
+    // This was taken directly from ADNS5050 driver...
+    // tSRW & tSRR. See page 15 of the ADNS5050 spec sheet.
+    // Technically, this is only necessary if the next operation is an SDIO
+    // read or write. This is not guaranteed to be the case.
+    // Honestly, this wait could probably be removed.
+    wait_us(1);
+
+    pmw3320_cs_deselect();
+
+    return byte;
+}
+
+void pmw3320_write_reg(uint8_t reg_addr, uint8_t data) {
+    pmw3320_cs_select();
+    pmw3320_serial_write(0b10000000 | reg_addr);
+    pmw3320_serial_write(data);
+    pmw3320_cs_deselect();
+}
+
+report_pmw3320_t pmw3320_read_burst(void) {
+    pmw3320_cs_select();
+
+    report_pmw3320_t data;
+    data.dx = 0;
+    data.dy = 0;
+
+    pmw3320_serial_write(REG_Motion_Burst);
+
+    uint8_t x = pmw3320_serial_read();
+    uint8_t y = pmw3320_serial_read();
+
+    // Burst mode returns a bunch of other shit that we don't really need.
```

Comment:
Sorry for obscene comment, it was in adns source and I forgot to change it...

According to [ADNS5050 datasheet](https://www.espruino.com/datasheets/ADNS5050.pdf), this is the data in the burst register:

> The sensor will return the data in the Delta_X, Delta_Y,
> SQUAL, Shutter_Upper, Shutter_Lower, Maximum_Pixel and Pixel_Sum. 

Not sure if info other than deltas really is helpful in getting mouse report.
[PMW3320 datasheet](https://www.epsglobal.com/Media-Library/EPSGlobal/Products/files/pixart/PMW3320DB-TYDU.pdf) don't have any information on this at all :(

Do you think that burst register on pmw3320 may also contain the delta_xy value?

Also, do you have any tips on how to use the data from delta_xy register? Should I concat its values as high part to the burst readings so the report becomes 12bit instead of 8bit?

Reply:
Tried a bit of debugging, seems like delta_xy (0x0c register) contain only these values: 0, 15, 240, 255, which are 0x00, 0x0F, 0xF0 and 0xFF. Not really sure what that would mean.

Third 8bits in motion_burst are giving me values from 70 to 95, which seems like surface quality, maybe?

Also found some interesting info in [ADNS2080 datasheet](https://pdf1.alldatasheet.com/datasheet-pdf/view/558621/PIXART/ADNS-2080.html). Looks like we should first enable 12-bit mode (probably by writing to the motion_control register), and then use delta_x/y and delta_xy registers instead of burst. Would be a bit slower, but more accurate, I guess?

Reply:
The problem with comparing datasheets like that, is that they're not the same.  Unfortunately, the 3320's datasheet is nowhere near as detailed as the 3360 or 3389.  Those have a lot more information about what is going on, how to use all the different registers, etc.  

For me, at least, the best method is just a raw dump of all the information to console, and then analyze that.

(unrelated, this makes me want to write a "3 pin SPI" driver, because this is the 3rd or 4th sensor to use this crapy SPI hackery)

Reply:
Well, I guess I can try to investigate and do something about 12-bit mode this weekend.

As for 3-wire SPI, yeah, that's what I thought about too, too much identical functions across different files for my taste :D Also it seems to be very common variation of usual SPI